### DISCUSSION TITLE
[Courtesy Issue] Nix package now generally available

### DISCUSSION THREAD

**Discussion by cmacrae** [asmvik/yabai#1244] [General]
👋 Hope you don't mind an issue being opened for this, but I figured it was a good way to notify you and an easy way for people to discover if they were looking for Nix support:

I submitted a [package for yabai to the nixpkgs repository](https://github.com/NixOS/nixpkgs/pull/86557) the other day. I'm happy to say, following the merge, it is now generally available on the `nixpkgs-unstable` channel 🎉 

```
$ nix-channel --list
darwin https://github.com/LnL7/nix-darwin/archive/master.tar.gz
nixpkgs https://nixos.org/channels/nixpkgs-unstable

$ nix-env -qaP --description yabai
nixpkgs.yabai  yabai-3.0.0  A tiling window manager for macOS based on binary space partitioning 
```

I'll aslo be submitting a Nix service module I've been using personally for a while to the [LnL7/nix-darwin](https://github.com/LnL7/nix-darwin) project. It lets you configure yabai like so:
```nix
{
  services.yabai = {
    enable = true;
    enableScriptingAddition = true;
    config = {
      focus_follows_mouse          = "autoraise";
      mouse_follows_focus          = "off";
      window_placement             = "second_child";
      window_opacity               = "off";
      window_opacity_duration      = "0.0";
      window_border                = "on";
      window_border_placement      = "inset";
      window_border_width          = 2;
      window_border_radius         = 3;
      active_window_border_topmost = "off";
      window_topmost               = "on";
      window_shadow                = "float";
      active_window_border_color   = "0xff5c7e81";
      normal_window_border_color   = "0xff505050";
      insert_window_border_color   = "0xffd75f5f";
      active_window_opacity        = "1.0";
      normal_window_opacity        = "1.0";
      split_ratio                  = "0.50";
      auto_balance                 = "on";
      mouse_modifier               = "fn";
      mouse_action1                = "move";
      mouse_action2                = "resize";
      layout                       = "bsp";
      top_padding                  = 36;
      bottom_padding               = 10;
      left_padding                 = 10;
      right_padding                = 10;
      window_gap                   = 10;
    };

    extraConfig = ''
        # rules
        yabai -m rule --add app='System Preferences' manage=off

        # Any other arbitrary config here
      '';
    };
}
```

Finally, I'm sure others here may be interested to know I also submitted a PR for spacebar and have a similar module ready to go, too. Once the package is generally available in the channels, I'll submit a PR for the module.

@koekeishiya @dominiklohmann Please feel free to simply close this issue - but if you'd like, I can update the documentation regarding the package (and perhaps, subsequently, the module - at a later date)

**Comment by dominiklohmann** [comment]
I think this would make a really good addition to the wiki.

Mind creating a draft of what you think needs to be mentioned? This can just be a rough outline / bullet points and I'll write a section from that.

**Comment by cmacrae** [comment]
Sure, I'll gladly draft something up :) Do you think it's perhaps worth waiting until I've got the module merged upstream (if it's accepted) in nix-darwin?  

That's the more comprehensive  part (in terms of use) of these contributions. I think if we updated the documentation to simply state a Nix package is available, it may leave some users wondering where an acompanying service is

**Comment by dominiklohmann** [comment]
> Do you think it's perhaps worth waiting until I've got the module merged upstream (if it's accepted) in nix-darwin?

People here are generally open to the idea of living on the edge and running nightly/unstable software, so I don't think it's a blocker.

**Comment by cmacrae** [comment]
Cool, thanks for weighing in here :) I'll take some time to draft something for the wiki a little later - I imagine it'll be fairly minimal

**Comment by cmacrae** [comment]
I've just submitted the service module PR to nix-darwin. LnL can be pretty snappy on this stuff, so we might even see a merge on this before I sit down and write the docs for the package anyway 🤞 
https://github.com/LnL7/nix-darwin/pull/196

**Comment by cmacrae** [comment]
It's now merged and generally available, short post on it here:
https://cmacr.ae/post/2020-05-13-yabai-module-in-nix-darwin-now-generally-available/

I'll write up a draft for docs tomorrow, as I should have some time :)

**Comment by cmacrae** [comment]
**Wiki Draft/Outline**
- A Nix package is also available on the `nixpkgs-unstable` channel
- A Nix service module is also available for users of [`nix-darwin`](https://github.com/LnL7/nix-darwin)
- Example configuration
```nix
{
  services.yabai = {
    enable = true;
    package = pkgs.yabai;
    enableScriptingAddition = true;
    config = {
      focus_follows_mouse          = "autoraise";
      mouse_follows_focus          = "off";
      window_placement             = "second_child";
      window_opacity               = "off";
      window_opacity_duration      = "0.0";
      window_topmost               = "on";
      window_shadow                = "float";
      active_window_opacity        = "1.0";
      normal_window_opacity        = "1.0";
      split_ratio                  = "0.50";
      auto_balance                 = "on";
      mouse_modifier               = "fn";
      mouse_action1                = "move";
      mouse_action2                = "resize";
      layout                       = "bsp";
      top_padding                  = 36;
      bottom_padding               = 10;
      left_padding                 = 10;
      right_padding                = 10;
      window_gap                   = 10;
    };

    extraConfig = ''
        # rules
        yabai -m rule --add app='System Preferences' manage=off

        # Any other arbitrary config here
      '';
    };
}
```

- Note: The `services.yabai.package` option must be specified and can simply be set to `pkgs.yabai`. There is no default option as it is currently distributed via the unstable channel
- Note: The `services.yabai.enableScriptingAddition` option only works is SIP is disabled

**Comment by cmacrae** [comment]
There you go @dominiklohmann :) Hopefully that's enough to go off, but please do let me know if there's stuff you feel I need to expand on

**Comment by breuerfelix** [comment]
thanks for the package! Is this currently bugged? :) I get some errors when trying to install from unstable right now on my M1

**Comment by ramnes** [comment]
Yep, looks like it depends on ghc-8.10.6 which isn't available for aarch64-darwin at the moment. Is there any workaround?

**Comment by venus-as-a-boy** [comment]
it's broken for me as well

```
./src/osax/payload.m:555:36: error: too many arguments to function call, expected 0, have 2
    return (uint64_t) objc_msgSend(space, @selector(spid));
                      ~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~~~~~~
/nix/store/f26dvgxlrs321kahfxy2bmdq8ap5wik3-libobjc-11.0.0/include/objc/message.h:63:1: note: 'objc_msgSend' declared here
objc_msgSend(void /* id self, SEL op, ... */ )
^
./src/osax/payload.m:560:60: error: too many arguments to function call, expected 0, have 3
    NSArray *spaces_for_display = (NSArray *) objc_msgSend(dock_spaces, @selector(spacesForDisplay:), display_uuid);
                                              ~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/nix/store/f26dvgxlrs321kahfxy2bmdq8ap5wik3-libobjc-11.0.0/include/objc/message.h:63:1: note: 'objc_msgSend' declared here
objc_msgSend(void /* id self, SEL op, ... */ )
^
./src/osax/payload.m:627:5: error: unknown register name '%rdi' in asm
    asm__call_move_space(source_space, dest_space, dest_display_uuid, dock_spaces, move_space_fp);
    ^
./src/osax/payload.m:606:148: note: expanded from macro 'asm__call_move_space'
        __asm__("movq %0, %%rdi;""movq %1, %%rsi;""movq %2, %%rdx;""movq %3, %%r13;""callq *%4;" : :"r"(v0), "r"(v1), "r"(v2), "r"(v3), "r"(func) :"%rdi", "%rsi", "%rdx", "%r13");
                                                                                                                                                   ^
./src/osax/payload.m:630:22: error: too many arguments to function call, expected 0, have 5
        objc_msgSend(dp_desktop_picture_manager, @selector(moveSpace:toDisplay:displayUUID:), source_space, dest_display_id, dest_display_uuid);
        ~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/nix/store/f26dvgxlrs321kahfxy2bmdq8ap5wik3-libobjc-11.0.0/include/objc/message.h:63:1: note: 'objc_msgSend' declared here
objc_msgSend(void /* id self, SEL op, ... */ )
^
./src/osax/payload.m:683:9: error: unknown register name '%rdi' in asm
        asm__call_add_space(new_space, display_space, add_space_fp);
        ^
./src/osax/payload.m:674:96: note: expanded from macro 'asm__call_add_space'
        __asm__("movq %0, %%rdi;""movq %1, %%r13;""callq *%2;" : :"r"(v0), "r"(v1), "r"(func) :"%rdi", "%r13");
                                                                                               ^
./src/osax/payload.m:694:40: error: too many arguments to function call, expected 0, have 3
        id source_space = objc_msgSend(dock_spaces, @selector(currentSpaceforDisplayUUID:), dest_display);
                          ~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/nix/store/f26dvgxlrs321kahfxy2bmdq8ap5wik3-libobjc-11.0.0/include/objc/message.h:63:1: note: 'objc_msgSend' declared here
objc_msgSend(void /* id self, SEL op, ... */ )
^
6 errors generated.
make: *** [makefile:24: src/osax/sa_loader.c] Error 1
builder for '/nix/store/g91lp3ms5mnkbmgq5vy3fmqyb4s7lbjs-yabai-3.3.4.drv' failed with exit code 2
```

**Comment by okkdev** [comment]
Any possibility of adding something like the nix darwin module to home-manager? 😄

**Comment by cmacrae** [comment]
> Any possibility of adding something like the nix darwin module to home-manager? 😄 

I don't use yabai at the moment. Why don't you try adding it, though? 😊 It should basically be a drop-in solution and could be a good learning experience

**Comment by okkdev** [comment]
> I don't use yabai at the moment. Why don't you try adding it, though? 😊 It should basically be a drop-in solution and could be a good learning experience

Hehe, I guess I'll have to try. Thanks for the quick reply! 😄

**Comment by bugeats** [comment]
Hey @cmacrae I'm a nix noob hoping that you can update the nix pkg to yabai `4.0.0`.

Also +1 for home-manager support.

**Comment by marcelmanz** [comment]
Having `4.0.0` on `nixpkgs` will be really cool :)

**Reply by marcelmanz** [reply]
Anyone knows why this error appears if I try to build from source:

```
warning: Git tree '/Users/m.manzanares/.nixpkgs' is dirty
warning: updating lock file '/Users/m.manzanares/.nixpkgs/flake.lock'
warning: Git tree '/Users/m.manzanares/.nixpkgs' is dirty
building the system configuration...
warning: Git tree '/Users/m.manzanares/.nixpkgs' is dirty
error: builder for '/nix/store/y9b7bdybpqsjlk3skvc65bxx0bjfqiav-yabai-4.0.0.drv' failed with exit code 2;
       last 10 log lines:
       > source root is source
       > patching sources
       > configuring
       > no configure script, doing nothing
       > building
       > build flags: SHELL=/nix/store/igw7h6k99kc03l2yp5j2p0rvzirp9b2h-bash-5.1-p16/bin/bash
       > rm -rf ./bin
       > xcrun clang ./src/osax/loader.m -shared -O2 -mmacosx-version-min=10.13 -arch x86_64 -o ./src/osax/loader -framework Foundation
       > /nix/store/igw7h6k99kc03l2yp5j2p0rvzirp9b2h-bash-5.1-p16/bin/bash: line 1: xcrun: command not found
       > make: *** [makefile:23: src/osax/loader_bin.c] Error 127
       For full logs, run 'nix log /nix/store/y9b7bdybpqsjlk3skvc65bxx0bjfqiav-yabai-4.0.0.drv'.
error: 1 dependencies of derivation '/nix/store/yaay7wzgdwgg4jjxkkfnd6vyx5h3vq0b-org.nixos.yabai.plist.drv' failed to build
error: 1 dependencies of derivation '/nix/store/imnpddy4xg8dkmmm71ak39wn3mj7szvc-system-applications.drv' failed to build
error: 1 dependencies of derivation '/nix/store/si9n1j5q1d7wcm90ar096xlckysx4pcw-yabai-sa-start.drv' failed to build
error: 1 dependencies of derivation '/nix/store/whwc8gprfwqb7fa7mh9hlpp0fq6m7bx8-darwin-system-22.05.20220506.d59dd43+darwin4.bb3baef.drv' failed to build
```

**Reply by emattiza** [reply]
Are xcode build tools on this system? xcrun command not found has me thinking that

**Reply by marcelmanz** [reply]
Yes.

`xcode-select -p` returns:
```
/Applications/Xcode.app/Contents/Developer
```

**Reply by daniels** [reply]
I've tried and failed to make yabai 4 build with nix (on M1). I don't think you can depend on Xcode `xcrun` being present. There is an `xcbuild` package that provides an alternate `xcrun` but it's not guaranteed to be compatible and is from an archived Facebook-project.

The furthest I've come is to remove the `xcrun` prefix, and instead provide `pkgs.llvmPackages_13.clang` as native build input. It will then complain about the platform as the default is to build a fat binary with both x86_64 and arm64e, so I've tried to replace all `-arch` flags in the `makefile` with `-arch arm64e`. That only takes me so far though, because in a later stage it complains that it can't build a fat binary where both alternatives are for the same platform. I don't know where this happens, as my modified `makefile` only provides a single `-arch` flag for each `clang` call, and I am in over my head.

Here's what I got so far – as an overlay/override to the existing yabai attribute:

```nix
        final: prev: {
          yabai = prev.yabai.overrideAttrs(attrs: {
            version = "4.0.0";
            src = prev.fetchFromGitHub {
              owner = "koekeishiya";
              repo = attrs.pname;
              rev = "v4.0.0";
              sha256 = "sha256-rllgvj9JxyYar/DTtMn5QNeBTdGkfwqDr7WT3MvHBGI=";
            };
            nativeBuildInputs = attrs.nativeBuildInputs ++ 
              [ prev.llvmPackages_13.clang];
            prePatch = ''
             substituteInPlace makefile \
                 --replace 'xcrun clang' clang
             substituteInPlace makefile --replace \
                 "-arch x86_64 -arch arm64e" "-arch arm64e"
             substituteInPlace makefile --replace \
                 "-arch x86_64 -arch arm64" "-arch arm64e"
             substituteInPlace makefile --replace \
                 "-arch x86_64" "-arch arm64e"
           ''; 
          });
```

**Reply by emptyflask** [reply]
I made an attempt at this before as well, and ran into a requirement for [ptrauth](https://github.com/koekeishiya/yabai/blob/master/src/osax/payload.m#L28-L33) which I tried to provide by including `darwin.xnu`, but didn't get any further.

**Reply by gil0mendes** [reply]
Hi people 👋🏾 

With my latest attempt, I got the following:

```
error: builder for '/nix/store/ap5ym7r6h1i5q0br1lw8qm1if8fsv8pz-yabai-3.3.10.drv' failed with exit code 2;
       last 10 log lines:
       >  ^
       > In file included from ./src/osax/payload.m:1:
       > In file included from /nix/store/9hpjsy91av3lx0pfbqsd7qcsnl5yhrjh-apple-framework-Foundation/Library/Frameworks/Foundation.framework/Headers/Foundation.h:6:
       > In file included from /nix/store/2xlw9hdlyj5rzwz14qjxfp742b4q4jbw-apple-framework-CoreFoundation/Library/Frameworks/CoreFoundation.framework/Headers/CoreFoundation.h:16:
       > /nix/store/c5jrszdq2hy3hj9rpyx3hk9gba60wd1d-Libsystem-1238.60.2/include/sys/types.h:95:9: error: unknown type name 'u_int64_t'
       > typedef u_int64_t               u_quad_t;       /* quads */
       >         ^
       > fatal error: too many errors emitted, stopping now [-ferror-limit=]
       > 20 errors generated.
       > make: *** [makefile:24: src/osax/loader_bin.c] Error 1
       For full logs, run 'nix log /nix/store/ap5ym7r6h1i5q0br1lw8qm1if8fsv8pz-yabai-3.3.10.drv'.
error: 1 dependencies of derivation '/nix/store/j9hqxfl913fy2q74kcdrpmh1xq86v8xk-org.nixos.yabai.plist.drv' failed to build
error: 1 dependencies of derivation '/nix/store/1yd5j15d204mjk08j0xzhpxz5canmhi7-system-applications.drv' failed to build
error: 1 dependencies of derivation '/nix/store/bcnc1nf4v837i8h1clwkhqmxvw5459p0-yabai-sa-start.drv' failed to build
error: 1 dependencies of derivation '/nix/store/f9hdi67x9dg8ya29s0v8wq3pmxg1n2nf-darwin-system-22.05.20220315.3eb07ee+darwin4.17fbc68.drv' failed to build
```

Any ideas? 🤔

**Reply by emptyflask** [reply]
> Hi people 👋🏾
> 
> With my latest attempt, I got the following:
> ```
> error: builder for '/nix/store/ap5ym7r6h1i5q0br1lw8qm1if8fsv8pz-yabai-3.3.10.drv' failed with exit code 2;
> ```

That's the version currently in nixpkgs-unstable. I'm able to build/run that with `nix-shell -p yabai` with no issues, but I'm on an M1 Macbook with OS 12.4. What are you doing?

**Reply by gil0mendes** [reply]
I don't noticed too, and that make me doubt if the overlay is working correctly. The overlay that I'm using is the following:

```nix
self: super: {
  yabai = super.yabai.overrideAttrs (old: {
    src = super.fetchFromGitHub {
      owner = "koekeishiya";
      repo = "yabai";
      rev = "v4.0.1";
      sha256 = "sha256-H1zMg+/VYaijuSDUpO6RAs/KLAAZNxhkfIC6CHk/xoI=";
    };
    nativeBuildInputs = old.nativeBuildInputs ++
      [ super.llvmPackages_13.clang ];
    prePatch = ''
      substituteInPlace makefile \
          --replace 'xcrun clang' clang
      substituteInPlace makefile --replace \
          "-arch x86_64 -arch arm64e" "-arch x86_64"
      substituteInPlace makefile --replace \
          "-arch x86_64 -arch arm64" "-arch x86_64"
      substituteInPlace makefile --replace \
          "-arch x86_64" "-arch x86_64"
    '';
  });
}
```

For context, I'm running in an x86_64 mac on version 12.3.

With the overlay above I got a new error, and looks like this is new code from version 4, but still saying version 3.3.10 🤔 

```
error: builder for '/nix/store/y0495gmn0c6w2k398hqakinfnf35x1d1-yabai-3.3.10.drv' failed with exit code 2;
       last 10 log lines:
       > rm -f ./src/osax/payload
       > rm -f ./src/osax/mach_loader
       > mkdir -p ./bin
       > clang src/manifest.m src/osax/loader_bin.c src/osax/payload_bin.c src/osax/mach_loader_bin.c -std=c99 -Wall -g -O0 -fvisibility=hidden -mmacosx-version-min=10.13 -fno-objc-arc -arch x86_64 -F/System/Library/PrivateFrameworks -framework Carbon -framework Cocoa -framework CoreServices -framework SkyLight -framework ScriptingBridge -o bin/yabai
       > In file included from src/manifest.m:72:
       > src/workspace.m:109:31: error: property 'safeAreaInsets' not found on object of type 'NSScreen *'
       >                 return screen.safeAreaInsets.top;
       >                               ^
       > 1 error generated.
       > make: *** [makefile:59: bin/yabai] Error 1
       For full logs, run 'nix log /nix/store/y0495gmn0c6w2k398hqakinfnf35x1d1-yabai-3.3.10.drv'.
error: 1 dependencies of derivation '/nix/store/ma5qks1c67g431dvw5wi80i6cw7ncqy8-org.nixos.yabai.plist.drv' failed to build
error: 1 dependencies of derivation '/nix/store/xdwqclq20kcp2cwflb39i1zjrxrxd978-system-applications.drv' failed to build
error: 1 dependencies of derivation '/nix/store/hcfvnxyld7jvij2rb8qs5ynx0irbca15-yabai-sa-start.drv' failed to build
error: 1 dependencies of derivation '/nix/store/ff1239j862nam2yjd7p3r8lvdsh2s71j-darwin-system-22.11.20220711.85deee6+darwin4.2f2bdf6.drv' failed to build
```

**Comment by IvarWithoutBones** [comment]
Hello, I just finished updating my yabai derivation to version 4.0.1 and thought people here might be interested. The scripting addition fails to load for me, but otherwise its perfectly stable. The version from nixpkgs (3.3.10) had a bunch of flickering problems for me on macOS 12.5, which are entirely fixed on the new update. Definitely great to see. I haven't had much luck with the scripting stuff on earlier versions either, so I'd recommend people to still give it a shot.

I do plan to PR the derivation to nixpkgs soon, I've listed myself as a maintainer on it as well so that we can hopefully get future updates merged a bit quicker :)

In the meantime, here it is from my dotfiles overlay: https://github.com/IvarWithoutBones/dotfiles/commit/f46f3bd475191fa60d4593b8170e9154958b0989

For reference, this is what I get when attempting to install SE, with no errors logged to the console:
<img width="370" alt="image" src="https://user-images.githubusercontent.com/41924494/185340501-799ced51-52b4-4246-96bd-2fe5ca2d49f9.png">

When attempting to load it:
```shell
~ > yabai --load-sa
yabai: scripting-addition either failed to load or could not inject payload into Dock.app! Error: -1708
```

`yabai --check-se` doesn't output anything and has a 0 exit code, even though SE clearly did not load.
```shell
~ > yabai --check-sa                                                           
~ > echo $?                                                                    
0
```

**Reply by asmvik** [reply]
The scripting-addition must be built as an arm64e binary for Apple Silicon, because it needs to be the same image format as the system application (Dock.app). Yabai itself should be built as an arm64 binary for Apple Silicon. For Intel based Macs, there shouldn't be a problem.

**Reply by IvarWithoutBones** [reply]
> For Intel based Macs, there shouldn't be a problem.

I do own an Intel based mac, yet I'm still hitting these problems.

**Reply by asmvik** [reply]
Can't really help you as I don't know anything about how nixpks work, but this issue is defniitely related to how nix works. Setting up yabai (with scripting-addition) outside of a nix environment has worked fine for years at this point.

**Reply by marcelmanz** [reply]
Yes I installed `yabai-v4.0.1`, via `brew`, on a macbook M1 Ultra, and it works fine.

**Reply by IvarWithoutBones** [reply]
> Can't really help you as I don't know anything about how nixpks work, but this issue is defniitely related to how nix works. Setting up yabai (with scripting-addition) outside of a nix environment has worked fine for years at this point.

Fair enough, thanks for taking the time to respond. I'm guessing the issue stems from yabai being compiled against different Apple SDK frameworks than the ones used in the user environment. I'll have to look into it a bit more.

Seeing how the scripting addition also didn't load for me with the yabai version in nixpkgs I'll open up a PR soon anyways, this can be fixed later.

> Yes I installed yabai-v4.0.1, via brew, on a macbook M1 Ultra, and it works fine.

Could you confirm the nix derivation I posted before doesn't work for you either?

**Comment by IvarWithoutBones** [comment]
Good news! I managed to get the scripting addition to load from the nix derivation, and opened the PR to nixpkgs :) https://github.com/NixOS/nixpkgs/pull/188322

For whatever reason neither `sudo yabai --install-sa` nor `yabai --load-sa` managed to execute the `mach_loader` binary in `/Library/ScriptingAdditions/yabai.osax/Contents/MacOS`, which caused the issues described before. When manually starting it with `sudo /Library/ScriptingAdditions/yabai.osax/Contents/MacOS/mach_loader` the scripting addition started, and all its features work flawlessly.

I figured this might be due to our sandbox messing up the path somehow, but `strings $(which yabai) | grep mach_loader` confirmed that wasn't the case. The file seems to have its permissions set correctly as well, and `codesign -dv /Library/ScriptingAdditions/yabai.osax/Contents/MacOS/mach_loader` claims it has ad-hoc codesigning enabled. 

Might be a hint that the installation procedure did correctly finish `scripting_addition_prepare_binaries`, but failed in `scripting_addition_restart_dock`?

Im not sure what could be causing the difference between the shell command and yabai's flag. Based on this yabai should execute the thing: https://github.com/koekeishiya/yabai/blob/6ef57aa2918ad18190901ea04fad37b0e51fb34a/src/sa.m#L412

 Something I find odd is that `sudo yabai --install-sa` does not print any errors to the console, just a notification. I cannot find any instance of only `notify` but not `warn` being called upon an error with the message I got. Perhaps it exits prematurely somewhere?

**Reply by asmvik** [reply]
The `--verbose` flag does not work with the `sa` flags, which is why the `debug` and/or `warn` is a no-op.

**Comment by richban** [comment]
@IvarWithoutBones  thanks for opening https://github.com/NixOS/nixpkgs/pull/188322 ! I tried to build your derivation on `aarch64-darwin` but it keeps complaining about `inherit (darwin.apple_sdk_11_0.frameworks) SkyLight;`

Is it because this SDK is not supported for Silicon?

**Reply by IvarWithoutBones** [reply]
I'm honestly not entirely sure. I'm pretty new to our darwin ecosystem, just got a macbook a few weeks ago 😅

I did notice cross compiling is broken from `x86_64-darwin` to `aarch64-darwin` is broken because it cannot find `Foundation.h`, see the comment in the `meta` section of the derivation in the PR. Marked it as broken on aarch64 because of that.

Not sure about native compilation on there though, I don't have a machine to test it with. Would you mind sharing the error you got? I don't see anything preventing aarch64 support in https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/darwin/apple-sdk-11.0/private-frameworks.nix#L17, so i assume that framework shouldn't be an issue. Version 11 of the SDK should be supported on aarch64 as far as I can tell.

**Reply by richban** [reply]
Don't understand why it complains about this path when it does exists. 

```bash
evaluating file '/nix/store/m7dvs21nxxw57489vj975r99rfkvhncw-source/system/darwin/yabai.nix'
error: attribute 'apple_sdk_11_0' missing

       at /nix/store/m7dvs21nxxw57489vj975r99rfkvhncw-source/system/overlays.nix:39:20:

           38|           inherit (darwin.apple_sdk.frameworks) Cocoa Carbon ScriptingBridge;
           39|           inherit (darwin.apple_sdk_11_0.private_frameworks) SkyLight;
             |                    ^
           40|         };
(use '--show-trace' to show detailed location information)
```

**Reply by IvarWithoutBones** [reply]
That line is supposed to be this:
```nix
inherit (darwin.apple_sdk_11_0.frameworks) SkyLight;
```
Might also need to update your nixpkgs revision, I believe the `apple_sdk_11_0` attribute wasn't added that long ago.

You can also try `nix build nixpkgs#darwin.apple_sdk_11_0.frameworks.SkyLight` to see if that can find it.

**Reply by IvarWithoutBones** [reply]
You can also do something along the lines of this this, i think:

```nix
yabai = (import (fetchurl {
  url = "url_of_my_branch.tar.gz";
}) { }).yabai;
```

Should be a bit more reliable since you'll just use the nixpkgs revision I tested this on.

**Reply by richban** [reply]
Indeed I had to update `nixpkgs` revisions. I tried to build but got this error:

```bash
error: builder for '/nix/store/fsw3b60qh11az4dpm7vjzpqvazz5aryp-yabai-4.0.2.drv' failed with exit code 2;
       last 10 log lines:
       > source root is source
       > patching sources
       > updateAutotoolsGnuConfigScriptsPhase
       > building
       > build flags: -j10 -l10 SHELL=/nix/store/jcwj04sh5kpwck3l9j9q0xgm9kj24mrz-bash-5.1-p16/bin/bash
       > rm -rf ./bin
       > xcrun clang ./src/osax/loader.m -shared -O2 -mmacosx-version-min=10.13 -arch x86_64 -o ./src/osax/loader -framework Foundation
       > error: unknown target CPU 'armv8.3-a+crypto+sha2+aes+crc+fp16+lse+simd+ras+rdm+rcpc'
       > note: valid target CPU values are: nocona, core2, penryn, bonnell, atom, silvermont, slm, goldmont, goldmont-plus, tremont, nehalem, corei7, westmere, sandybridge, corei7-avx, ivybridge, core-avx-i, haswell, core-avx2, broadwell, skylake, skylake-avx512, skx, cascadelake, cooperlake, cannonlake, icelake-client, icelake-server, tigerlake, knl, knm, k8, athlon64, athlon-fx, opteron, k8-sse3, athlon64-sse3, opteron-sse3, amdfam10, barcelona, btver1, btver2, bdver1, bdver2, bdver3, bdver4, znver1, znver2, x86-64
       > make: *** [makefile:23: src/osax/mach_loader_bin.c] Error 1
       For full logs, run 'nix log /nix/store/fsw3b60qh11az4dpm7vjzpqvazz5aryp-yabai-4.0.2.drv'.
```

**Reply by IvarWithoutBones** [reply]
Ah, any chance you're on an M1 mac? I've only tested this on x86 as I dont own an aarch64 machine. I posted a patch that might fix this in a [comment on my PR](https://github.com/NixOS/nixpkgs/pull/188322#issuecomment-1232046481), would you mind testing it out?

Its a shame cross compiling from `x86_64-darwin` to `aarch64-darwin` doesn't work, I'd love to help out more.

**Reply by richban** [reply]
Same error message as: https://github.com/NixOS/nixpkgs/pull/188322#issuecomment-1232356421

**Reply by IvarWithoutBones** [reply]
@richban I switched to the binary release for `aarch64-darwin` on my PR as I couldn't get the from-source build to work. It should at least work now on `aarch64-darwin` :)

**Reply by richban** [reply]
I have managed to build it based on your PR but for some reason I am having issues with loading the scripting addition I am getting this error:

```bash
evaluating file '/nix/store/ds82bads7cj70b7n6n8mpbxgi00daqka-source/modules/launchd/lib.nix'
error: attribute 'loadScriptingAddition' missing

       at /nix/store/kmdwpi8r10qqymggkhmxq2siwslvxh7k-source/system/darwin/wm.nix:19:32:

           18|     path = [ pkgs.yabai config.environment.systemPath ];
           19|     command = "/usr/bin/sudo ${pkgs.yabai.loadScriptingAddition}";
             |                                ^
           20|     serviceConfig = {
(use '--show-trace' to show detailed location information)
> ./result/sw/bin/darwin-rebuild switch --flake .#darwinWorkM1
```

This is the file: https://github.com/richban/dotfiles/blob/master/system/darwin/wm.nix#L10-L24 and added yabai as an overlay https://github.com/richban/dotfiles/blob/master/system/overlays.nix#L31-L43

**Reply by IvarWithoutBones** [reply]
The hack to load the scripting addition unfortunately doesn't work on `aarch64-darwin`, which is why its only available on `x86_64-darwin`. See the last few comments on the PR, someone is attempting to debug this.

**Comment by richban** [comment]
How do you restart `yabai` when installed via `nix` or from source?

**Reply by IvarWithoutBones** [reply]
I use the `nix-darwin` module which starts it as a `launchctl` service. With that, I restart yabai + skhd with the following:


```console
$ launchctl kickstart -k gui/${UID}/org.nixos.yabai && launchctl kickstart -k gui/${UID}/org.nixos.skhd
```

For a from-source build it depends on how you start yabai in the first place, although I would imagine it is usually the same.

**Reply by richban** [reply]
Thanks! `launchctl` works for me!

**Comment by IvarWithoutBones** [comment]
I've updated yabai to 5.0.1 in https://github.com/NixOS/nixpkgs/pull/203504. This fixes the issues we had with the scripting addition both on Intel and M1 machines, the redesigned `--load-sa` flag works perfectly. I've explained how to apply the PR locally in a comment on the PR, in case anyone wants to try it out before it gets merged.

**Comment by IvarWithoutBones** [comment]
Updated the PR to 5.0.2.
